stl/inc/bit

```diff
@@ -94,20 +98,120 @@ _NODISCARD constexpr _Ty rotr(const _Ty _Val, const int _Rotation) noexcept {
-_NODISCARD constexpr int countl_zero(const _Ty _Val) noexcept {
+#pragma warning(push)
+#pragma warning(disable : 4640)
+
+template <class _Ty>
+_NODISCARD constexpr int _Popcount_helper(_Ty _Val) {
     constexpr int _Digits = numeric_limits<_Ty>::digits;
-    if (_Val == 0) {
+    // we static cast these bit patterns in order to truncate
+    // them to the correct size
+    _Val = static_cast<_Ty>(_Val - ((_Val >> 1) & static_cast<_Ty>(0x5555555555555555ull)));
+    _Val = static_cast<_Ty>(
+        (_Val & static_cast<_Ty>(0x3333333333333333ull)) + ((_Val >> 2) & static_cast<_Ty>(0x3333333333333333ull)));
+    _Val = static_cast<_Ty>((_Val + (_Val >> 4)) & static_cast<_Ty>(0x0F0F0F0F0F0F0F0Full));
+    for (int _ShiftDigits = 8; _ShiftDigits < _Digits; _ShiftDigits <<= 1) {
+        _Val = static_cast<_Ty>(_Val + static_cast<_Ty>(_Val >> _ShiftDigits));
+    }
+    // we want the bottom "slot" that's big enough to store _Digits
+    return static_cast<int>(_Val & static_cast<_Ty>(_Digits + _Digits - 1));
+}
+
+#if defined(_M_IX86) || defined(_M_X64)
+
+extern "C" {
+__MACHINEX86_X64(unsigned int __lzcnt(unsigned int))
+__MACHINEX86_X64(unsigned short __lzcnt16(unsigned short))
+__MACHINEX64(unsigned __int64 __lzcnt64(unsigned __int64))
+__MACHINEX86_X64(unsigned int __popcnt(unsigned int))
+__MACHINEX86_X64(unsigned short __popcnt16(unsigned short))
+__MACHINEX64(unsigned __int64 __popcnt64(unsigned __int64))
+extern int __isa_available;
+}
+
+template <class _Ty>
+_NODISCARD inline int _Checked_x86_x86_64_countl_zero(_Ty _Val) {
+    const bool _Have_lzcnt = __isa_available >= __ISA_AVAILABLE_AVX;
+    constexpr int _Digits  = numeric_limits<_Ty>::digits;
+    // lzcnt (when it doesn't fall back to bsr) is defined correctly for zero
+    if (!_Have_lzcnt && _Val == 0) {
         return _Digits;
     }
+    // we use lzcnt (actually bsr if lzcnt is not supported) for all cases now we know
+    // we're not zero. We can do this because lzcnt and bsr share the same instruction
+    // encoding.
+    if constexpr (_Digits <= 16) {
+        return static_cast<int>(__lzcnt16(_Val) - (16 - _Digits));
+    } else if constexpr (_Digits <= 32) {
+        return static_cast<int>(__lzcnt(_Val) - (32 - _Digits));
+    } else {
+#ifdef _M_IX86
+        unsigned int _High = _Val >> 32;
+        unsigned int _Low  = static_cast<unsigned int>(_Val);
+        int _Result        = _Checked_x86_x86_64_countl_zero(_High);
+        _Result += !_High ? _Checked_x86_x86_64_countl_zero(_Low) : 0;
+        return _Result;
+#else
+        return static_cast<int>(__lzcnt64(_Val) - (64 - _Digits));
+#endif
+    }
+    // note we don't need to call a fallback here because
+    // all supported intel processors at least have bsr/bsf
+}
+
+template <class _Ty>
+_NODISCARD inline int _Checked_x86_x86_64_popcount(_Ty _Val) {
+    const bool _Have_popcnt = __isa_available >= __ISA_AVAILABLE_SSE42;

```

Comment:
Can't check POPCNT bit directly?

Reply:
There is no popcnt bit, timing is annoying to add it. don't loose that much anyway.

Reply:
to clarify there is a popcnt bit in __cpuid, but those magic statics were not gunna work out.